src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Listeners/ServiceBusListener.cs

```diff
@@ -124,20 +125,45 @@ public async Task StopAsync(CancellationToken cancellationToken)
                 throw new InvalidOperationException("The listener has not yet been started or has already been stopped.");
             }
 
-            // cancel our token source to signal any in progress
-            // ProcessMessageAsync invocations to cancel
-            _cancellationTokenSource.Cancel();
-
-            if (_receiver != null && _receiver.IsValueCreated)
+            if (_isStopping)
             {
-                await Receiver.CloseAsync();
-                _receiver = CreateMessageReceiver();
+                throw new InvalidOperationException("The listener is currently stopping");
             }
-            if (_clientEntity != null)
+
+            // Unregister* methods stop new messages from being processed while allowing in-flight messages to complete.
+            // As the amount of time functions are allowed to complete processing varies by SKU, we specify max timespan
+            // as the amount of time Service Bus SDK should wait for in-flight messages to complete procesing after 
+            // unregistering the message handler so that functions have as long as the host continues to run time to complete.
+            if (_singleDispatch)
             {
-                await _clientEntity.CloseAsync();
-                _clientEntity = null;
+                _isStopping = true;
+
+                if (_isSessionsEnabled)
+                {
+                    if (_clientEntity != null)
+                    {
+                        if (_clientEntity is QueueClient queueClient)
+                        {
+                            await queueClient.UnregisterSessionHandlerAsync(TimeSpan.MaxValue);

```

Comment:
what does` TimeSpan.MaxValue` parameter mean?

Reply:
This is how long UnregisterSessionHandlerAsync will allow messages to drain (finish processing) before timing out the operation.  As the amount of time functions host will allow functions to complete processing before shutting down depends on the SKU, this ensures we don't prematurely terminate in-flight messages without having to dynamically set a timeout value depending on the SKU (which is unknown by the extension code).  If triggered functions exceed the SLA for the SKU, the host will shutdown anyway which closes the Service Bus connection.